src/imgproc/imgproc.h

```diff
@@ -246,27 +246,90 @@ class imgproc {
 
     auto hist0 = cache.put_buffer(256 << 2, Types::S32_t | Types::C1_t);
 
-    for (; i < 256; ++i)
-      hist0->i32[i] = 0;
+    // for (; i < 256; ++i)
+    //   hist0->i32[i] = 0;
     for (i = 0; i < size; ++i) {
-      ++hist0->i32[src->u8[i]];
+      //++hist0->i32[src->u8[i]];
+      hist0->i32[src->u8[i]] = ++i;
     }
 
     prev = hist0->i32[0];
     for (i = 1; i < 256; ++i) {
-      prev = hist0->i32[i] += prev;
+      // prev = hist0->i32[i] += prev;
+      hist0->i32[i] += prev;
+      prev = hist0->i32[i];
+      std::cout << hist0->i32[i] << std::endl;
     }
 
     auto norm = 255 / size;
     for (i = 0; i < size; ++i) {
+      // dst->u8[i] = (unsigned char)(hist0->i32[src->u8[i]] * norm + 0.5) | 0;
       dst->u8[i] = (unsigned char)(hist0->i32[src->u8[i]] * norm + 0.5) | 0;
     }
     // to do: buffer should be put down the head...
   };
+
+  void equalize_histogram_internal2(matrix_t *src, matrix_t *dst) {
+    auto w = src->cols, h = src->rows;
+
+    dst->resize(w, h, src->channel);
+
+    /* max_val must be 255 if we use unsigned char as input, this should be
+       changed if we use another input type (short, int...)
+     */
+    int max_val = 255;
```

Comment:
as written in the comment this is applyable only for unsigned char types. For this reason we should implement a templated class that may accept different types.